Add CookieChangeListener base class and make WebCookieCache a CookieChangeListener

RupinMittal opened the pull request:
#### 08f56b30694ca5a973f159ad51c86a6b294fcbce
<pre>
Add CookieChangeListener base class and make WebCookieCache a CookieChangeListener
<a href="https://bugs.webkit.org/show_bug.cgi?id=259710">https://bugs.webkit.org/show_bug.cgi?id=259710</a>

Reviewed by Chris Dumez.

This patch is a step towards implementing the notifications upon
changes in cookies part of the new Cookie Store API. We would like
the CookieStore to listen for changes in cookies. So we add a
CookieChangeListener abstract base class which requires subclasses
to define what should happen when a cookie is added or deleted.

Currently, we have a WebCookieCache in the Web Process which listens
to updates in cookies (per host) from the Network Process. It sends
an IPC to NetworkConnectionToWebProcess to subscribe to cookie-update
notifications. When a change happends, NetworkConnectionToWebProcess
sends an IPC with the update to WebCookieaJar, which notifies the
WebCookieCache.

The new design is that: WebCookieCache is a subclass of
CookieChangeListener (so it specifies what to do on cookie add/delete).
As before, when it receives a request to obtain cookies for a host and
it&apos;s the first request for this host, WebCookieCache will subscribe to
listen to updates. Except, instead of directly sending an IPC to the
NetworkProcess, it will register itself as a listener with the
WebCookieJar. The WebCookieJar will maintain a map from hosts to their
listeners. The WebCookieJar will send an IPC to the Network Process
to subscribe the listener. On updates, the WebCookieJar will be sent
an IPC and it will notify the relevant listeners for that host. When
the WebCookieCache wants to unsubscribe to notifications for a certain
host, it will tell the WebCookieJar, which will send the IPC to the
Network Process to unsubscribe the listener. It will also update it&apos;s
map accordingly.

A future patch will make the CookieStore a CookieChangeListener too.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/CookieChangeListener.h: Copied from Source/WebKit/WebProcess/WebPage/WebCookieCache.h.
(WebCore::CookieChangeListener::~CookieChangeListener):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::addChangeListener):
(WebCore::CookieJar::removeChangeListener):
* Source/WebCore/loader/CookieJar.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::domCookiesForHost):
(WebKit::NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications):
(WebKit::NetworkConnectionToWebProcess::unsubscribeFromCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::cookiesAdded):
(WebKit::NetworkProcessConnection::cookiesDeleted):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::cookiesForDOM):
(WebKit::WebCookieCache::cookiesAdded):
(WebKit::WebCookieCache::clear):
(WebKit::WebCookieCache::clearForHost):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookiesAdded):
(WebKit::WebCookieJar::cookiesDeleted):
(WebKit::WebCookieJar::addChangeListener):
(WebKit::WebCookieJar::removeChangeListener):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/266570@main">https://commits.webkit.org/266570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b767c7ad3d425b24a18a21574dfbed02bfc44524

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14445 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15871 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13395 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14217 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16955 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14528 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16070 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14880 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11982 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16584 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12163 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12743 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19762 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13242 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12907 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16110 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13450 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11314 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12730 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3432 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17065 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13294 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 51;
+	objectVersion = 54;
```
Please revert this change.

cdumez commented on Source/WebCore/loader/CookieJar.h:
```diff
@@ -67,6 +67,9 @@ class WEBCORE_EXPORT CookieJar : public RefCounted<CookieJar> {
     virtual void getCookiesAsync(Document&, const URL&, const CookieStoreGetOptions&, CompletionHandler<void(std::optional<Vector<Cookie>>&&)>&&) const;
     virtual void setCookieAsync(Document&, const URL&, const Cookie&, CompletionHandler<void(bool)>&&) const;
 
+    virtual void addCookieChangeListener(const String&, const CookieChangeListener&);
```
Please don't omit the first parameter name as it is not obvious what it is.

cdumez commented on Source/WebCore/loader/CookieJar.h:
```diff
@@ -67,6 +67,9 @@ class WEBCORE_EXPORT CookieJar : public RefCounted<CookieJar> {
     virtual void setCookieAsync(Document&, const URL&, const Cookie&, CompletionHandler<void(bool)>&&) const;
 
+    virtual void addCookieChangeListener(const String&, const CookieChangeListener&);
+    virtual void removeCookieChangeListener(const String&);
```
ditto.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -200,4 +200,12 @@ void CookieJar::setCookieAsync(Document&, const URL&, const Cookie&, CompletionH
+{
 }
+
+void CookieJar::removeCookieChangeListener(const String&)
```
This needs to take the CookieChangeListener too, just like the `add` version.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
```diff
@@ -67,34 +68,28 @@ void WebCookieCache::setCookiesFromDOM(const URL& firstParty, const SameSiteInfo
 }
 
-void WebCookieCache::cookiesAdded(const String& host, const Vector<Cookie>& cookies)
+void WebCookieCache::allCookiesDeleted()
```
Can we restore the previous function ordering so that the diff is not so hard to read?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,32 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+        return;
+
+    for (auto& cookie : cookies)
+        cookieChangeListener->addCookie(host, WTFMove(cookie));
```
Why are we now passing the cookies to the observer one by one instead of passing the full vector like we used to do. Seems less efficient.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,32 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+        return;
+
+    for (auto& cookie : cookies)
+        cookieChangeListener->deleteCookie(host, WTFMove(cookie));
```
ditto.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +329,18 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+void WebCookieJar::removeCookieChangeListener(const String& host)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::UnsubscribeFromCookieChangeNotifications(HashSet<String> { host }), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
+    m_cookieChangeListeners.remove(host);
```
That's wrong, there can be several listeners for the same host. You need to pass the listener as parameter.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
```diff
@@ -67,6 +76,7 @@ class WebCookieJar final : public WebCore::CookieJar {
     bool isEligibleForCache(WebFrame&, const URL& firstPartyForCookies, const URL& resourceURL) const;
 
     mutable WebCookieCache m_cache;
+    HashMap<String, WeakPtr<WebCore::CookieChangeListener>> m_cookieChangeListeners;
```
This should be a `HashMap<String, WeakHashSet<WebCore::CookieChangeListener>>` since there can be more than one listener per host.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -641,8 +641,8 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
 #if HAVE(CGIMAGESOURCE_DISABLE_HARDWARE_DECODING)
-        OSStatus ok = CGImageSourceDisableHardwareDecoding();
-        ASSERT_UNUSED(ok, ok == noErr);
+//        OSStatus ok = CGImageSourceDisableHardwareDecoding();
```
I know you need to do this locally but please omit from PRs (or mark as draft).

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -43,14 +43,15 @@ messages -> NetworkConnectionToWebProcess LegacyReceiver {
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
-    UnsubscribeFromCookieChangeNotifications(HashSet<String> hosts)
+    SubscribeToCookieChangeNotifications(String host) AllowedWhenWaitingForSyncReply
+    UnsubscribeFromCookieChangeNotifications(HashSet<String> hosts) AllowedWhenWaitingForSyncReply
```
Seems we should pass a single host now instead of a HashSet. You never call with more than one host now I believe.

cdumez commented on Source/WebCore/loader/CookieChangeListener.h:
```diff
@@ -0,0 +1,46 @@
+
+class CookieChangeListener : public CanMakeWeakPtr<CookieChangeListener> {
+public:
+    virtual void addCookie(const String& host, Cookie&&) = 0;
```
Ideally this would be cookiesAdded() and take a Vector of Cookies.

cdumez commented on Source/WebCore/loader/CookieChangeListener.h:
```diff
@@ -0,0 +1,46 @@
+class CookieChangeListener : public CanMakeWeakPtr<CookieChangeListener> {
+public:
+    virtual void addCookie(const String& host, Cookie&&) = 0;
+    virtual void deleteCookie(const String& host, Cookie&&) = 0;
```
Ideally this would be cookiesDeleted() and take a Vector of Cookies.

cdumez commented on Source/WebCore/loader/CookieJar.h:
```diff
@@ -67,6 +67,11 @@ class WEBCORE_EXPORT CookieJar : public RefCounted<CookieJar> {
     virtual void setCookieAsync(Document&, const URL&, const Cookie&, CompletionHandler<void(bool)>&&) const;
 
+#if HAVE(COOKIE_CHANGE_LISTENER_API)
+    virtual void addCookieChangeListener(const String& host, const CookieChangeListener&);
```
Since this is on a `CookieJar` class, I think we can get rid of `Cookie` in the function names and use `addChangeListener() / removeChangeListener()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
```diff
@@ -25,18 +25,22 @@
 #include <WebCore/CookieJar.h>
 #include <WebCore/SameSiteInfo.h>
 #include <wtf/HashSet.h>
+#include <wtf/Vector.h>
```
Why are we including Vector.h and WTFString here? Would including `<wtf/Forward.h>` suffice?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
```diff
@@ -45,8 +49,10 @@ class WebCookieCache {
-    void cookiesAdded(const String& host, const Vector<WebCore::Cookie>&);
-    void cookiesDeleted(const String& host, const Vector<WebCore::Cookie>&);
+    // CookieChangeListener
+    void cookiesAdded(const String& host, const Vector<WebCore::Cookie>&) final;
```
These two can probably be private now.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,26 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+void WebCookieJar::cookiesAdded(const String& host, Vector<WebCore::Cookie>&& cookies)
 {
-    m_cache.cookiesAdded(host, cookies);
+    auto cookieChangeListenersIter = m_cookieChangeListeners.find(host);
```
We can probably just call this `it`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,26 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+void WebCookieJar::cookiesDeleted(const String& host, Vector<WebCore::Cookie>&& cookies)
 {
-    m_cache.cookiesDeleted(host, cookies);
+    auto cookieChangeListenersIter = m_cookieChangeListeners.find(host);
```
We can probably just call this `it`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,28 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+#if HAVE(COOKIE_CHANGE_LISTENER_API)
+void WebCookieJar::addCookieChangeListener(const String& host, const WebCore::CookieChangeListener& cookieChangeListener)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::SubscribeToCookieChangeNotifications(host), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
```
We should only send the IPC if this is the first listener for this host and if the listener is actually new.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,28 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::UnsubscribeFromCookieChangeNotifications(host), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
+
+    auto listenersIter = m_cookieChangeListeners.find(host);
```
We can call this `it`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,28 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+        return;
+
+    listenersIter->value.remove(cookieChangeListener);
+    if (!listenersIter->value.computeSize())
```
I'd rather we call `isEmptyIgnoringNullReferences()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,28 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+
+void WebCookieJar::removeCookieChangeListener(const String& host, const WebCore::CookieChangeListener& cookieChangeListener)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::UnsubscribeFromCookieChangeNotifications(host), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
```
We should only send this IPC if there is no longer any listener for the host.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
```diff
@@ -67,6 +76,7 @@ class WebCookieJar final : public WebCore::CookieJar {
     bool isEligibleForCache(WebFrame&, const URL& firstPartyForCookies, const URL& resourceURL) const;
 
     mutable WebCookieCache m_cache;
+    HashMap<String, WeakHashSet<WebCore::CookieChangeListener>> m_cookieChangeListeners;
```
`m_changeListeners` should suffice given that this is on a CookieJar.

cdumez commented on Source/WebCore/loader/CookieChangeListener.h:
```diff
@@ -0,0 +1,47 @@
+#include <wtf/WeakPtr.h>
+
+namespace WTF {
+class String;
```
Either include `<wtf/Forward.h>` or don't do nothing. This forward include attempt clearly does nothing since the code below uses `String`, not `WTF::String`.

cdumez commented on Source/WebCore/loader/CookieChangeListener.h:
```diff
@@ -0,0 +1,47 @@
+
+#pragma once
+
+#include <wtf/Vector.h>
```
Not needed, we should just include <wtf/Forward.h>

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,26 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+    if (it == m_changeListeners.end())
+        return;
+
+    auto cookieChangeListenersHashSet = it->value;
```
Instead of doing your own WeakHashSet copy here, you could also just call WeakHashSet::forEach() which iterates safely on a copy for you:
```
it->value.forEach([&](auto& listener) {
    listener.cookiesDeleted(host, cookies);
});
```

Also, I think we can just use `listener` instead of the very long `cookieChangeListener`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -177,14 +178,26 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
+    if (it == m_changeListeners.end())
+        return;
+
+    auto cookieChangeListenersHashSet = it->value;
```
Same comments as below for using forEach() and shorter listener name.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,33 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+void WebCookieJar::addChangeListener(const String& host, const WebCore::CookieChangeListener& cookieChangeListener)
+{
+    auto it = m_changeListeners.find(host);
+    if (it != m_changeListeners.end() && it->value.contains(cookieChangeListener))
```
This is *still* not correct. You're now avoiding sending the subscribe IPC when this listener is already registered. However, you're *still* unnecessarily sending the subscribe IPC if there is already a listener for this host.

RupinMittal replied:
Should be fixed now!

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,33 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+    if (it != m_changeListeners.end() && it->value.contains(cookieChangeListener))
+        return;
+
+    m_changeListeners.ensure(host, [] {
```
`auto addResult = ...`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,33 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+    m_changeListeners.ensure(host, [] {
+        return WeakHashSet<CookieChangeListener> { };
+    }).iterator->value.add(cookieChangeListener);
+
```
```
auto& listenersForHost = m_changeListeners.add(host, WeakHashSet<CookieChangeListener> { });
auto addResult = listenersForHost.add(cookieChangeListener);
if (listenersForHost.size() != 1 || !addResult.isNewEntry)
    return;
```
is what I would do.

RupinMittal replied:
If there are already listeners for this host in `m_changeListeners`, wouldn't 
`auto& listenersForHost = m_changeListeners.add(host, WeakHashSet<CookieChangeListener> { });`
remove them because it will replace the WeakHashSet which contains those listeners with an empty WeakHashSet?

cdumez replied:
No, add() doesn’t replace, unlike the set() function.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,33 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::SubscribeToCookieChangeNotifications(host), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
+}
+
+void WebCookieJar::removeChangeListener(const String& host, const WebCore::CookieChangeListener& cookieChangeListener)
```
This function looks correct though.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -310,4 +323,34 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+void WebCookieJar::addChangeListener(const String& host, const WebCore::CookieChangeListener& listener)
+{
+    auto it = m_changeListeners.find(host);
+    if (it != m_changeListeners.end() && it->value.contains(listener))
```
We no longer need the ` && it->value.contains(listener)` check.

RupinMittal replied:
What if there are listeners for this host already but this listener is not one of them? Then without the 
` && it->value.contains(listener)` check, this if statement will be true, and we will return without adding this listener to `m_changeListeners` right?

cdumez replied:
Sorry, you need to get rid of this whole find() and contains logic. The code I suggested below should suffice. You ended up using both your code and my suggested code instead of replacing.

RupinMittal replied:
Oh got it. Fixed now!

cdumez reviewed (approved):
LGTM with minor fix.

RupinMittal committed:
Add CookieChangeListener base class and make WebCookieCache a CookieChangeListener
https://bugs.webkit.org/show_bug.cgi?id=259710

Reviewed by Chris Dumez.

This patch is a step towards implementing the notifications upon
changes in cookies part of the new Cookie Store API. We would like
the CookieStore to listen for changes in cookies. So we add a
CookieChangeListener abstract base class which requires subclasses
to define what should happen when a cookie is added or deleted.

Currently, we have a WebCookieCache in the Web Process which listens
to updates in cookies (per host) from the Network Process. It sends
an IPC to NetworkConnectionToWebProcess to subscribe to cookie-update
notifications. When a change happends, NetworkConnectionToWebProcess
sends an IPC with the update to WebCookieaJar, which notifies the
WebCookieCache.

The new design is that: WebCookieCache is a subclass of
CookieChangeListener (so it specifies what to do on cookie add/delete).
As before, when it receives a request to obtain cookies for a host and
it's the first request for this host, WebCookieCache will subscribe to
listen to updates. Except, instead of directly sending an IPC to the
NetworkProcess, it will register itself as a listener with the
WebCookieJar. The WebCookieJar will maintain a map from hosts to their
listeners. The WebCookieJar will send an IPC to the Network Process
to subscribe the listener. On updates, the WebCookieJar will be sent
an IPC and it will notify the relevant listeners for that host. When
the WebCookieCache wants to unsubscribe to notifications for a certain
host, it will tell the WebCookieJar, which will send the IPC to the
Network Process to unsubscribe the listener. It will also update it's
map accordingly.

A future patch will make the CookieStore a CookieChangeListener too.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/CookieChangeListener.h: Copied from Source/WebKit/WebProcess/WebPage/WebCookieCache.h.
(WebCore::CookieChangeListener::~CookieChangeListener):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::addChangeListener):
(WebCore::CookieJar::removeChangeListener):
* Source/WebCore/loader/CookieJar.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::domCookiesForHost):
(WebKit::NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications):
(WebKit::NetworkConnectionToWebProcess::unsubscribeFromCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::cookiesAdded):
(WebKit::NetworkProcessConnection::cookiesDeleted):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::cookiesForDOM):
(WebKit::WebCookieCache::cookiesAdded):
(WebKit::WebCookieCache::clear):
(WebKit::WebCookieCache::clearForHost):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookiesAdded):
(WebKit::WebCookieJar::cookiesDeleted):
(WebKit::WebCookieJar::addChangeListener):
(WebKit::WebCookieJar::removeChangeListener):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: https://commits.webkit.org/266570@main
